README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rameritrade

<!-- badges: start -->

![CRAN
Version](https://www.r-pkg.org/badges/version/rameritrade?color=green)
![Dev Version](https://img.shields.io/badge/github-0.1.5-blue.svg)
![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/rameritrade)
<!-- badges: end -->

R package for the TD Ameritrade API, facilitating authentication,
trading, price requests, account balances, positions, order history,
option chains, and more. A user will need a TD Brokerage account and TD
Ameritrade developer app. Read the article [Trade on
TD](https://exploringfinance.github.io/posts/2020-10-17-trade-on-td-ameritrade-with-r/)
for a full example of logging in and executing a trade, or see the
instructions below.

## Introduction

TD Ameritrade is one of many trading platforms that offer a trade API.
Others include Alpaca, RobinHood, InteractiveBrokers, and eTrade. Alpaca
and Robinhood offer great capabilities, and there are existing R
packages for both. Unfortunately, they do not offer the full
capabilities of a major brokerage firm such as IRAs, multiple accounts,
etc. InteractiveBrokers requires the IB workstation to be open and
active which can make it hard to build automated trading strategies
using tools like CRON jobs. Using the TD API you can fully automate
trade execution across multiple accounts and multiple logins, assuming
you have access and permission to do so. This can be a great way to
dollar cost average into the market for an IRA!

The initial authentication requires a few manual steps, but once initial
authorization is granted, all API calls can be fully automated without
needing to manually log in again. Additionally, because of the use of
tokens and a middle layer App, user name and passwords never need to be
entered into the R code. This can help protect the security of accounts
assuming tokens are stored securely.

### Disclosure

This software is in no way affiliated, endorsed, or approved by TD
Ameritrade or any of its affiliates. It comes with absolutely no
warranty and should not be used in actual trading unless the user can
read and understand the source code. The functions within this package
have been tested under basic scenarios. There may be bugs or issues that
could prevent a user from executing trades or canceling trades. It is
also possible trades could be submitted in error. The user will use this
package at their own risk.

Due to the Charles Schwab acquisition of TD Ameritrade, the fate of the
TD API is unknown. There have been multiple indications that Schwab is
excited about the technology at TD Ameritrade and will retain as much as
possible. That being said, this package could stop working at any point
if the acquisition results in the termination of the API.

Please heed the following warning for the `td_placeOrder` function.
WARNING: TRADES THAT ARE SUCCESSFULLY ENTERED WILL BE SUBMITTED
IMMEDIATELY THERE IS NO REVIEW PROCESS. THE `td_placeOrder` FUNCTION HAS
HUNDREDS OF POTENTIAL COMBINATIONS AND ONLY A HANDFUL HAVE BEEN TESTED.
TD AMERITRADE HAS THEIR OWN ERROR HANDLING BUT IF A SUCCESSFUL
COMBINATION IS ENTERED IT COULD BE EXECUTED IMMEDIATELY. DOUBLE CHECK
ALL ENTRIES BEFORE SUBMITTING. IT IS STRONGLY RECOMMENDED TO TEST THE
DESIRED ORDER FIRST. THREE POTENTIAL OPTIONS ARE:

1.  Enter trades outside of normal market trading hours and check the TD
    website to ensure proper entry
2.  Use limit orders with limit prices far outside the current bid/ask
3.  Enter very small quantities that won’t put much capital at risk

FOR OPTIONS 1 AND 2, BE SURE TO CANCEL ORDERS USING `td_cancelOrder` OR
THROUGH THE TD WEBSITE

## Installation

You can install rameritrade using:

``` r
# Available on CRAN
install.packages("rameritrade")

# Install development version - fixes for cronR and price history
# install.packages("devtools")
devtools::install_github("exploringfinance/rameritrade")
```

## Authentication

Initial authorization to a TD Brokerage account requires a 3 step
authentication process. Once initial authorization is achieved, tokens
can be used to maintain access indefinitely. Below is a detailed summary
of the entire process followed by code demonstrating the 3 step process.
More can be found at the [TD authentication
FAQ](https://developer.tdameritrade.com/content/authentication-faq) or
the [Authentication
Guide](https://developer.tdameritrade.com/content/getting-started#registerApp).
Details are also provided within the functions.

1.  Register an API Developer account with [TD Ameritrade
    Developer](https://developer.tdameritrade.com/).
2.  Create an app under My Apps. The app serves as a middle layer
    between the brokerage account and API.
3.  Identify the Consumer Key provided by TD (essentially an API Key).
4.  Under Edit App, create a Callback URL. This can be relatively simple
    (for example: `https://YourAppName`).
5.  Pass the Consumer Key and Callback URL to `td_auth_loginURL` to
    generate a URL specific to the app for user log in.
6.  Visit the URL in a web browser and log in to a TD Brokerage account,
    granting the app access to the user account.
7.  When “Allow” is clicked, it will redirect to a blank page or
    potentially an error page stating “This site can’t be reached”. This
    indicates a successful log in. The URL of this page is the
    authorization code (`https://YourAppName/?code=AUTHORIZATIONCODE`).
8.  Feed the Consumer Key, Callback URL, and authorization code into
    `td_auth_refreshToken` to get a Refresh Token.
9.  The Refresh Token is valid for 90 days so be sure to store it
    somewhere safe. The Refresh Token is the only component needed from
    then on for account access. However, if your token expires or is
    lost, you can always follow steps 5-8 above.
10. The Refresh Token is used to generate an Access Token using
    `td_auth_accessToken` which gives account access for 30 minutes.
11. The most recent Access Token is stored by default into Options.
    Passing it into the functions is optional unless accessing multiple
    accounts.
12. To reset the Refresh Token as it approaches expiration, you can pass
    a valid Refresh Token into `td_auth_refreshToken`.

Please note: TD has indicated they prefer infrequent token generation
and will take action on excessive tokens being generated

#### Terminology

-   Authorization Code: generated from using a TD Brokerage to log into
    the `td_auth_loginURL`.
-   Refresh Token: generated using the Authorization Code or an existing
    Refresh Token and is used to create access tokens. Refresh token is
    valid for 90 days.
-   Access Token: generated using the Refresh Token and creates the
    connection to the API. Valid for 30 minutes.

## Authentication Example

The `td_auth_loginURL` is used to gain initial access to the API. Once a
Refresh Token is generated using an authorization code, manual log in
will not be required unless the Refresh Token expires. This can be
avoided by passing a valid Refresh Token to `td_auth_refreshToken`.

``` r
# --------- Step 1 -----------
# Register an App with TD Ameritrade Developer, create a Callback URL, and get a Consumer Key.
# The callback URL can be anything (for example: https://myTDapp).
# Use the td_auth_loginURL to generate an app specific URL. See the TD Authentication FAQ for issues.

callbackURL = 'https://myTDapp'
consumerKey = 'TD_CONSUMER_KEY'

rameritrade::td_auth_loginURL(consumerKey, callbackURL)
# "https://auth.tdameritrade.com/auth?response_type=code&redirect_uri=https://myTDapp&client_id=consumerKey%40AMER.OAUTHAP"

# Visit the URL above to see a TD login screen. Log in with a TD Brokerage account to grant the app access. 


# --------- Step 2 -----------
# A successful log in to the URL from Step 1 will result in a blank page once "Allow" is clicked. 
# The URL of this blank page is the Authorization Code. 
# The blank page may indicate "This site can't be reached". The URL is still a valid Authorization Code.
# Feed the Authorization Code URL into td_auth_refreshToken to get a Refresh Token.

authCode = 'https://myTDapp/?code=AUTHORIZATIONCODE' # This could be over 1,000 alpha numeric characters
refreshToken = rameritrade::td_auth_refreshToken(consumerKey, callbackURL, authCode)
# "Successful Refresh Token Generated"

# Save the Refresh Token to a safe location so it can be retrieved as needed. It will be valid for 90 days.
saveRDS(refreshToken,'/secure/location/')


# --------- Step 3 -----------
# Use the Refresh Token to get an Access Token
# The function will return an Access Token and also store it for use as a default token in Options

refreshToken = readRDS('/secure/location/')
accessToken = rameritrade::td_auth_accessToken(consumerKey, refreshToken)
# "Successful Login. Token has been stored and will be valid for 30 minutes"

# Authentication has been completed. Other functions can now be used.


# --------- Step 4 (when needed) -----------
# The Refresh Token should be reset before it expires after 90 days. 
# TD indicates they do look for frequent Refresh Token generation. This should be used conservatively. 

refreshToken = readRDS('/secure/location/')
refreshToken = rameritrade::td_auth_refreshToken(consumerKey, codeToken = refreshToken) # Callback URL is not required
# "Successful Refresh Token Generated"
saveRDS(refreshToken,'/secure/location/')
```

## Get Account Data

Use the `td_accountData` to get current account data that includes
balances, positions, and current day orders.

``` r
library(rameritrade)

refreshToken = readRDS('/secure/location/')
consumerKey = 'TD_CONSUMER_KEY'
accessToken = rameritrade::td_auth_accessToken(consumerKey, refreshToken)

actDF = td_accountData()
str(actDF)
# List of 3
# $ balances : tibble [2 × 40] (S3: tbl_df/tbl/data.frame)
# ..$ accountId                       : chr [1:2] "1234" "1234"
# ..$ type                            : chr [1:2] "CASH" "MARGIN"
# ..$ roundTrips                      : int [1:2] 0 0
# ..$ isDayTrader                     : logi [1:2] FALSE TRUE
# ..$ isClosingOnlyRestricted         : logi [1:2] FALSE FALSE
# ..$ accruedInterest                 : num [1:2] 0 0
# ..$ cashBalance                     : num [1:2] 0 9.76
# ..$ cashReceipts                    : num [1:2] 0 0
# ..$ longOptionMarketValue           : num [1:2] 0 0
# ..$ liquidationValue                : num [1:2] 33009 35505

actList = td_accountData('list')
str(actList)
# List of 3
# $ balances :List of 2
# ..$ :List of 1
# .. ..$ securitiesAccount:List of 8
# .. .. ..$ type                   : chr "CASH"
# .. .. ..$ accountId              : chr "1234"
# .. .. ..$ roundTrips             : int 0
# .. .. ..$ isDayTrader            : logi FALSE
# .. .. ..$ isClosingOnlyRestricted: logi FALSE
```

## Get Pricing Data

Use the `price` functions to get quotes or historical pricing. Quotes
will be real-time if the account has access to real-time quotes.

``` r
library(rameritrade)

refreshToken = readRDS('/secure/location/')
consumerKey = 'TD_CONSUMER_KEY'
accessToken = rameritrade::td_auth_accessToken(refreshToken, consumerKey)

### Quote data
SP500Qt = rameritrade::td_priceQuote(c('SPY', 'IVV', 'VOO'))
str(SP500Qt)

# 'data.frame': 3 obs. of  48 variables:
# $ assetType                         : chr  "ETF" "ETF" "ETF"
# $ assetMainType                     : chr  "EQUITY" "EQUITY" "EQUITY"
# $ cusip                             : chr  "78462F103" "464287200" "922908363"
# $ assetSubType                      : chr  "ETF" "ETF" "ETF"
# $ symbol                            : chr  "SPY" "IVV" "VOO"
# $ description                       : chr  "SPDR S&P 500" "iShares Core S&P 500 ETF" "Vanguard S&P 500 ETF"
# $ bidPrice                          : num  331 332 305


# Historical Data
SP500H = rameritrade::td_priceHistory(c(c('SPY','IVV','VOO')))
head(SP500H)
# A tibble: 6 x 8
# ticker date       date_time            open  high   low close   volume
# <chr>  <date>     <dttm>              <dbl> <dbl> <dbl> <dbl>    <int>
# 1 SPY    2020-08-18 2020-08-18 01:00:00  338.  339.  337.  339. 38733908
# 2 SPY    2020-08-19 2020-08-19 01:00:00  339.  340.  337.  337. 68054244
# 3 SPY    2020-08-20 2020-08-20 01:00:00  335.  339.  335.  338. 42207826
# 4 SPY    2020-08-21 2020-08-21 01:00:00  338.  340.  338.  339. 55106628
# 5 SPY    2020-08-24 2020-08-24 01:00:00  342.  343   339.  343. 48588662
# 6 SPY    2020-08-25 2020-08-25 01:00:00  344.  344.  342.  344. 38463381


# Time series data
# History is only available back to a certain time depending on frequency
rameritrade::td_priceHistory('AAPL', startDate = '2020-09-01', freq='5min')
# # A tibble: 2,424 x 8
# ticker date       date_time            open  high   low close volume
# <chr>  <date>     <dttm>              <dbl> <dbl> <dbl> <dbl>  <int>
# 1 AAPL   2020-09-01 2020-09-01 07:00:00  132.  132.  132.  132. 203104
# 2 AAPL   2020-09-01 2020-09-01 07:05:00  132.  132.  132.  132.  85287
# 3 AAPL   2020-09-01 2020-09-01 07:10:00  132.  132   131.  131.  93742
# 4 AAPL   2020-09-01 2020-09-01 07:15:00  131.  132.  131.  132.  63895
# 5 AAPL   2020-09-01 2020-09-01 07:20:00  132.  132.  131.  131.  26498

```

## Placing Trades

Order entry offers hundreds of potential combinations. It is strongly
recommended to submit trades outside market hours first to test the
trade entries. You can confirm proper entry on the TD website before
canceling. See the [order sample
guide](https://developer.tdameritrade.com/content/place-order-samples)
for more examples. Please note, `td_placeOrder` only allows for single
order entry and will not support some of the complex examples in the
guide.

``` r
library(rameritrade)

# Set Access Token using a valid Refresh Token
refreshToken = readRDS('/secure/location/')
consumerKey = 'TD_CONSUMER_KEY'
accessToken = rameritrade::td_auth_accessToken(refreshToken, consumerKey)
accountNumber = 1234567890

# Market Order
Ord0 = rameritrade::td_placeOrder(accountNumber,
                                  ticker = 'PSLV',
                                  quantity = 1,
                                  instruction = 'BUY')
rameritrade::td_cancelOrder(Ord0$orderId, accountNumber)
# [1] "Order Cancelled"



# Good till cancelled stop limit INCORRECT ENTRY
Ordr1 = rameritrade::td_placeOrder(accountNumber = accountNumber,
                                  ticker = 'SCHB',
                                  quantity = 1,
                                  instruction = 'buy',
                                  duration = 'good_till_cancel',
                                  orderType = 'stop_limit',
                                  limitPrice = 50,
                                  stopPrice = 49)
# Error: 400 - The stop price must be above the current ask for buy stop orders 
#        and below the bid for sell stop orders.



# Good till Cancelled Stop Limit Order correct entry
Ordr1 = rameritrade::td_placeOrder(accountNumber = accountNumber,
                                   ticker = 'SCHB',
                                   quantity = 1,
                                   instruction = 'buy',
                                   duration = 'good_till_cancel',
                                   orderType = 'stop_limit',
                                   limitPrice = 86,
                                   stopPrice = 85)
rameritrade::td_cancelOrder(Ordr1$orderId, accountNumber)
# [1] "Order Cancelled"



# Trailing Stop Order
Ordr2 = rameritrade::td_placeOrder(accountNumber = accountNumber,
                                   ticker = 'SPY',
                                   quantity = 1,
                                   instruction = 'sell',
                                   orderType = 'trailing_stop',
                                   stopPriceBasis = 'BID',
                                   stopPriceType = 'percent',
                                   stopPriceOffset = 10)
rameritrade::td_cancelOrder(Ordr2$orderId,accountNumber)
# [1] "Order Cancelled"

# Option Order
Ord3 = rameritrade::td_placeOrder(accountNumber = accountNumber,
                                  ticker = 'SLV_091820P24.5',
                                  quantity = 1,
                                  instruction = 'BUY_TO_OPEN',
                                  duration = 'Day',
                                  orderType = 'LIMIT',
                                  limitPrice = .02,
                                  assetType = 'OPTION')
rameritrade::td_cancelOrder(Ord3$orderId, accountNumber)
# [1] "Order Cancelled"
```

## Option Chains

You can pull entire option chains for individual securities.

``` r
library(rameritrade)
consumerKey = 'TD_CONSUMER_KEY'

refreshToken1 = readRDS('/secure/location/')
accessToken = rameritrade::td_auth_accessToken(refreshToken, consumerKey)

# Pull all SPY chains for 6 months with 12 strikes above and below current market
SPY = td_optionChain('SPY',
                     strikes = 12,
                     endDate = Sys.Date() + 180)

# This returns a list of two data frames
str(SPY$underlying)
# tibble [1 × 23] (S3: tbl_df/tbl/data.frame)
# $ symbol           : chr "SPY"
# $ description      : chr "SPDR S&P 500"
# $ change           : num 2.52
# $ percentChange    : num 0.76
# $ close            : num 332
# $ quoteTime        : num 1.6e+12
# $ tradeTime        : num 1.6e+12
# $ bid              : num 334
# $ ask              : num 334
# $ last             : num 335
# $ mark             : num 334
# $ markChange       : num 1.53
# $ markPercentChange: num 0.46
# $ bidSize          : int 300
# $ askSize          : int 100
# $ highPrice        : num 338
# $ lowPrice         : num 333
# $ openPrice        : num 333
# $ totalVolume      : int 101506148
# $ exchangeName     : chr "PAC"
# $ fiftyTwoWeekHigh : num 359
# $ fiftyTwoWeekLow  : num 218
# $ delayed          : logi TRUE

str(SPY$fullChain)
# $ putCall               : chr [1:552] "PUT" "PUT" "PUT" "PUT" ...
# $ symbol                : chr [1:552] "SPY_093020P329" "SPY_093020P330" "SPY_093020P331" "SPY_093020P332" ...
# $ description           : chr [1:552] "SPY Sep 30 2020 329 Put (Quarterly)" "SPY Sep 30 2020 330 Put (Quarterly)" "SPY Sep 30 2020 331 Put (Quarterly)" "SPY Sep 30 2020 332 Put (Quarterly)" ...
# $ exchangeName          : chr [1:552] "OPR" "OPR" "OPR" "OPR" ...
# $ bid                   : num [1:552] 0 0 0.01 0.01 0.04 0.3 1.15 2.02 3.14 4.1 ...
# $ ask                   : num [1:552] 0.01 0.01 0.02 0.02 0.05 0.38 1.25 2.33 3.24 4.61 ...
# $ last                  : num [1:552] 0.01 0.02 0.01 0.02 0.05 0.32 1.22 2.35 3.02 4.3 ...
# $ mark                  : num [1:552] 0.01 0.01 0.02 0.02 0.05 0.34 1.2 2.17 3.19 4.36 ...
# $ bidSize               : int [1:552] 0 0 6739 1457 390 40 10 10 10 15 ...
# $ askSize               : int [1:552] 4927 3498 6062 3177 10 15 10 141 10 150 ...
# $ bidAskSize            : chr [1:552] "0X4927" "0X3498" "6739X6062" "1457X3177" ...
# $ lastSize              : int [1:552] 0 0 0 0 0 0 0 0 0 0 ...
# $ highPrice             : num [1:552] 0.33 0.49 0.67 1 1.4 1.93 2.58 3.2 4.1 5 ...

NEWS.md
# rameritrade 0.1.5

## Release Notes and News

### rameritrade 0.1.5 In development

Fix for `td_priceHistory` to correct for scipen = 999 requirement.
Fix within `ram_checkRefresh` for cronR.
Change maintainer.

### rameritrade 0.1.4 - 10/7/2020
Approved by CRAN: 2020-10-15 with commit d7ee4af.

Fixed `td_accountData` orders and added '' marks
around 'TD Ameritrade' API in description and name.

### rameritrade 0.1.3 - 10/1/2020

Before CRAN approval, modified naming convention for functions 
and consolidated multiple functions into single functions such 
as account data, authentication, and pricing.

### rameritrade 0.1.2 - 9/29/2020

Added `` around example URL in ReadMe

### rameritrade 0.1.1 - 9/26/2020

Removed git files from build

### rameritrade 0.1.0 - 9/26/2020

Initial release includes basic functionality and API calls. Authentication,
Placing Orders, Canceling Orders, and pulling account details are all included
with the initial release.

Updating User Preferences and complex order entry are not included.

Disclosure: 
This software is in no way affiliated, endorsed, or approved by TD
Ameritrade or any of its affiliates. It comes with absolutely no warranty and
should not be used in actual trading unless the user can read and understand the
source code. The functions within this package have been tested under basic
scenarios. There may be bugs or issues that could prevent a user from executing
trades or canceling trades. It is also possible trades could be submitted in
error. The user will use this package at their own risk.

tests/testthat.R
library(testthat)
library(rameritrade)

test_check("rameritrade")


R/orders.R
#' Get Details for a Single Order
#'
#' Pass an order ID and Account number to get details such as status, quantity,
#' ticker, executions (if applicable), account, etc.
#'
#' @param orderId A valid TD Ameritrade Order ID
#' @param accountNumber The TD brokerage account number associated with the
#'   Access Token
#' @inheritParams td_accountData
#'
#' @return list of order details
#' @export
#'
#' @examples
#' \dontrun{
#'
#' # Get stored refresh token
#' refreshToken = readRDS('/secure/location/')
#'
#' # generate a new access token
#' accessToken = td_auth_accessToken(refreshToken, 'consumerKey')
#'
#' # Get order details for a single order
#' # Passing Access Token is optional once it's been set
#' td_orderDetail(orderId = 123456789, accountNumber = 987654321)
#'
#' }
td_orderDetail = function(orderId, accountNumber, accessToken=NULL) {
  
  # Get access token from options if one is not passed
  accessToken = ram_accessToken(accessToken)
  
  # Get Order Details
  orderURL = paste0('https://api.tdameritrade.com/v1/accounts/',accountNumber,'/orders/',orderId)
  orderDetails = httr::GET(orderURL,ram_headers(accessToken))
  
  # Confirm status code of 200
  ram_status(orderDetails)
  
  # Return order details in list form
  httr::content(orderDetails)

}

#' Cancel an Open Order
#' 
#' Pass an Order ID and Account number to cancel an existing open order
#'
#' @inheritParams td_orderDetail
#'
#' @return order API URL. Message confirming cancellation
#' @export
#'
#' @examples
#'  \dontrun{
#' 
#' td_cancelOrder(orderId = 123456789, accountNumber = 987654321)
#' 
#' }
td_cancelOrder =  function(orderId,accountNumber,accessToken=NULL){
  
  # Get access token from options if one is not passed
  accessToken = ram_accessToken(accessToken)
  
  # Create Order URL and then DELETE order
  orderURL = paste0('https://api.tdameritrade.com/v1/accounts/',accountNumber,'/orders/',orderId)
  orderCancel = httr::DELETE(orderURL,ram_headers(accessToken))
  
  # Confirm status code of 200
  ram_status(orderCancel,
             '. Make sure the order ID is for an open order and the account number is correct.')
  
  # Print confirmation message and display location
  print('Order Cancelled')
  orderCancel$url
  
}


#' Search for orders by date
#'
#' Search for orders associated with a TD account over the previous 60 days. The
#' result is a list of three objects: 
#' \enumerate{ 
#' \item jsonlite formatted extract of all orders 
#' \item all entered orders with details 
#' \item a data frame of all executed orders with the executions }
#'
#' @inheritParams td_orderDetail
#' @param startDate Orders from a certain date with. Format yyyy-mm-dd. TD
#'   indicates there is a 60 day max, but this limit may not always apply
#' @param endDate Filter orders that occurred before a certain date. Format
#'   yyyy-mm-dd
#' @param maxResults the max results to return in the query
#' @param orderStatus search by order status (ACCEPTED, FILLED, EXPIRED,
#'   CANCELED, REJECTED, etc)
#'
#' @return a list of three objects: a jsonlite formatted extract of all orders,
#'   all entered orders with details, a data frame of all executed orders with
#'   the executions
#' @export
#'
#' @examples
#' \dontrun{
#'
#' # Get all orders run over the last 50 days (up to 500)
#' td_orderSearch(accountNumber = 987654321, 
#'              startDate = Sys.Date()-days(50),
#'              maxResult = 500, orderStatus = 'FILLED')
#'
#' }
td_orderSearch = function(accountNumber, startDate = Sys.Date()-30, endDate = Sys.Date(),
                        maxResults = 50, orderStatus = '', accessToken = NULL){
  
  # Bind variables to quiet warning
  accountId <- orderId <- instrument.symbol <- instruction <- total_qty <- NULL
  quantity <- duration <- orderType <- instrument.cusip <- enteredTime <- NULL
  
  # Get access token from options if one is not passed
  accessToken = ram_accessToken(accessToken)
  
  # Construct URL for GET request
  searchURL = paste0('https://api.tdameritrade.com/v1/orders?accountId=',accountNumber,
                     '&maxResults=',maxResults,'&status=',orderStatus,
                     '&fromEnteredTime=',startDate,'&toEnteredTime=',endDate)
  
  # Make GET request
  searchOrders = httr::GET(searchURL,ram_headers(accessToken),encode='json')
 
  
  # Confirm status code of 200
  ram_status(searchOrders)
  
  # Extract content from GET request
  jsonOrder <- httr::content(searchOrders, as = "text",encoding = 'UTF-8')
  jsonOrder <- jsonlite::fromJSON(jsonOrder)
  
  
  OrdrExecFinal=NULL
  OrderEnterFinal=NULL
    # Run a loop for each order within the account
    UnqOrdrs = httr::content(searchOrders)
    for(ords in 1:length(UnqOrdrs)) {
      
      # Get the high level order details
      OrdrDet = UnqOrdrs[[ords]]
      OrdrDet$orderLegCollection = NULL
      OrdrDet$orderActivityCollection = NULL
      OrdrDet = data.frame(OrdrDet) %>% dplyr::rename(total_qty=quantity)
      
      # Get the Entry details and merge with order details
      OrdrEnter = UnqOrdrs[[ords]]
      OrdrEnter = dplyr::bind_rows(lapply(OrdrEnter$orderLegCollection,data.frame))
      OrdrEnter = merge(OrdrEnter,OrdrDet)
      OrderEnterFinal = dplyr::bind_rows(OrderEnterFinal,OrdrEnter)
      
      # Get execution details when available
      OrdrExec = UnqOrdrs[[ords]]
      OrdrExec = dplyr::bind_rows(lapply(OrdrExec$orderActivityCollection,data.frame))
      OrdrEntDet = dplyr::select(OrdrEnter,accountId,orderId,instrument.symbol,instruction,total_qty,duration,orderType,instrument.cusip,
                                 enteredTime)
      OrdrExecAll = merge(OrdrEntDet,OrdrExec)
      OrdrExecFinal = dplyr::bind_rows(OrdrExecFinal,OrdrExecAll)
    }
  
    # Combine all three outputs into a single list
    orderOutput = list(enteredOrders = dplyr::as_tibble(OrderEnterFinal),
                       executedOrders = dplyr::as_tibble(OrdrExecFinal),
                       allOrderJSON = dplyr::as_tibble(jsonOrder))
    
    orderOutput
}  




#' Place Order for a specific account
#'
#' Place trades through the TD Ameritrade API using a range of parameters
#'
#' A valid account and access token must be passed. An access token will be
#' passed by default when \code{\link{td_auth_accessToken}} is executed
#' successfully and the token has not expired, which occurs after 30 minutes.
#' Only equities and options can be traded at this time. This function is built
#' to allow a single trade submission. More complex trades can be executed
#' through the API, but a custom function or submission will need to be
#' constructed. To build more custom trading strategies, reference the
#' \href{https://developer.tdameritrade.com/account-access/apis}{TD Ameritrade
#' API Instructions} or the
#' \href{https://developer.tdameritrade.com/content/place-order-samples}{order
#' sample guide}. A full list of the input parameters and details can be found
#' at the links above. Please note that in rare cases, the documentation may not
#' be accurate in the API section, so the Order Sample guide is a better
#' reference. TEST ALL ORDERS FIRST WITH SMALL DOLLAR AMOUNTS!!!
#'
#' Four parameters are required for submission: ticker, instruction, quantity,
#' and account number associated with the Access Token. The following parameters
#' default: session - NORMAL, duration - DAY, asset type - EQUITY, and order
#' type - MARKET
#'
#' @section Warning: TRADES THAT ARE SUCCESSFULLY ENTERED WILL BE SUBMITTED
#'   IMMEDIATELY THERE IS NO REVIEW PROCESS. THIS FUNCTION HAS HUNDREDS OF
#'   POTENTIAL COMBINATIONS AND ONLY A HANDFUL HAVE BEEN TESTED. IT IS STRONGLY
#'   RECOMMENDED TO TEST THE DESIRED ORDER ON A VERY SMALL QUANTITY WITH LITTLE
#'   MONEY AT STAKE. ANOTHER OPTION IS TO USE LIMIT ORDERS FAR FROM THE CURRENT
#'   PRICE. TD AMERITRADE HAS THEIR OWN ERROR HANDLING BUT IF A SUCCESSFUL
#'   COMBINATION IS ENTERED IT COULD BE EXECUTED IMMEDIATELY. DOUBLE CHECK ALL
#'   ENTRIES BEFORE SUBMITTING.
#'
#'
#' @inheritParams td_orderDetail
#' @param ticker a valid Equity/ETF or option. If needed, use td_symbolDetail to
#'   confirm. This should be a ticker/symbol, not a CUSIP
#' @param quantity the number of shares to be bought or sold. Must be an
#'   integer.
#' @param instruction Equity instructions include 'BUY', 'SELL', 'BUY_TO_COVER',
#'   or 'SELL_SHORT'. Options instructions include 'BUY_TO_OPEN',
#'   'BUY_TO_CLOSE', 'SELL_TO_OPEN', or 'SELL_TO_CLOSE'
#' @param orderType MARKET, LIMIT (requiring limitPrice), STOP (requiring
#'   stopPrice), STOP_LIMIT, TRAILING_STOP (requiring stopPriceBasis,
#'   stopPriceType, stopPriceOffset)
#' @param limitPrice the limit price for a LIMIT or STOP_LIMIT order
#' @param stopPrice the stop price for a STOP or STOP_LIMIT order
#' @param assetType EQUITY or OPTION. No other asset types are available at this
#'   time. EQUITY is the default.
#' @param session NORMAL for normal market hours, AM or PM for extended market
#'   hours
#' @param duration how long will the trade stay open without a fill: DAY,
#'   GOOD_UNTIL_CANCEL, FILL_OR_KILL
#' @param stopPriceBasis LAST, BID, or ASK which is the basis for a STOP,
#'   STOP_LIMIT, or TRAILING_STOP
#' @param stopPriceType the link to the stopPriceBasis. VALUE for dollar
#'   difference or PERCENT for a percentage offset from the price basis
#' @param stopPriceOffset an integer that indicates the offset used for the
#'   stopPriceType, 10 and PERCENT is a 10 percent offset from the current price
#'   basis. 5 and VALUE is a 5 dollar offset from the current price basis
#'
#' @return the trade id, account id, and other order details
#' @export
#'
#' @examples
#' \dontrun{
#'
#' # Get stored refresh token
#' refreshToken = readRDS('/secure/location/')
#'
#' # generate a new access token
#' accessToken = td_auth_accessToken(refreshToken, 'consumerKey')
#'
#' # Set Account Number
#' accountNumber = 1234567890
#'
#' # Standard market buy order
#' # Every order must have at least these 4 paramters
#' td_placeOrder(accountNumber = accountNumber,
#'             ticker = 'AAPL',
#'             quantity = 1,
#'             instruction = 'buy')
#'
#' # Stop limit order - good until canceled
#' td_placeOrder(accountNumber = accountNumber,
#'             ticker = 'AAPL',
#'             quantity = 1,
#'             instruction = 'sell',
#'             duration = 'good_till_cancel',
#'             orderType = 'stop_limit',
#'             limitPrice = 98,
#'             stopPrice = 100)
#'
#' # Trailing Stop Order
#' td_placeOrder(accountNumber = accountNumber,
#'             ticker='AAPL',
#'             quantity = 1,
#'             instruction='sell',
#'             orderType = 'trailing_stop',
#'             stopPriceBasis = 'BID',
#'             stopPriceType = 'percent',
#'             stopPriceOffset = 10)
#'
#' # Option Order with a limit price
#' td_placeOrder(accountNumber = accountNumber,
#'             ticker = 'SLV_091820P24.5',
#'             quantity = 1,
#'             instruction = 'BUY_TO_OPEN',
#'             duration = 'Day',
#'             orderType = 'LIMIT',
#'             limitPrice = .02,
#'             assetType = 'OPTION')
#'
#' }
#'
#' 
td_placeOrder = function(accountNumber, ticker, quantity, instruction,
                       orderType = 'MARKET', limitPrice = NULL, stopPrice = NULL,
                       assetType = c('EQUITY','OPTION'), session='NORMAL', duration='DAY',
                       stopPriceBasis = NULL, stopPriceType = NULL, stopPriceOffset = NULL,
                       accessToken = NULL) {
  
  # Get access token from options if one is not passed
  accessToken = ram_accessToken(accessToken)
  
  # Check symbol and asset type
  if (missing(assetType)) assetType ='EQUITY'
  
  # Set URL specific to account
  orderURL = paste0('https://api.tdameritrade.com/v1/accounts/',accountNumber,'/orders')
  
  # Put order details in a list
  orderList = list(orderType = orderType,
                   complexOrderStrategyType = 'NONE',
                   session = session,
                   duration = duration,
                   price = limitPrice,
                   stopPrice = stopPrice,
                   orderStrategyType = 'SINGLE',
                   stopPriceLinkBasis = toupper(stopPriceBasis),
                   stopPriceLinkType = toupper(stopPriceType),
                   stopPriceOffset = stopPriceOffset,
                   orderLegCollection = list(list(
                     instruction = instruction,
                     quantity = quantity,
                     instrument = list(
                       symbol = ticker,
                       assetType = assetType
                     )
                   ))
  )
  
  # Post order to TD
  postOrder = httr::POST(orderURL,ram_headers(accessToken),body=orderList,encode='json')
  
  # Confirm status code of 201
  ram_status(postOrder)
  
  # Collect Order Details
  orderDet = postOrder$headers
  orderOutput = data.frame(
    accountNumber = gsub('/orders/.*','',gsub('.*accounts/','',orderDet$location)),
    orderId = gsub('.*orders/','',orderDet$location),
    status_code = postOrder$status_code,
    date = orderDet$date,
    location = orderDet$location
  )
  
  # Return Order Output
  orderOutput
}







